### PR TITLE
Emit details of the contents of literal-value uniforms (#1029)

### PR DESCRIPTION
**Pull request by amakin-siru**

### CODE REVIEW

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 1]
```diff
@@ -176,6 +200,11 @@ public void visitVariablesDeclaration(VariablesDeclaration variablesDeclaration)
     visit(baseType);
     out.append(" ");
     boolean first = true;
+
+    // If the uniform value supplier provides values for this variable, emit a comment that
+    // contains the values.
+    Map<String, String> values = new HashMap<>();
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:207]: I think something like this code would be good:

```
    if (uniformValues.isPresent() && variablesDeclaration.getBaseType().hasQualifier(TypeQualifier.UNIFORM)) {
      for (VariableDeclInfo vdi : variablesDeclaration.getDeclInfos()) {
        final Optional<List<Number>> values = uniformValues.get().getValues(vdi.getName());
        values.ifPresent(item -> {
          out.append("// Contents of ")
              .append(vdi.getName())
              .append(": [")
              .append(item.stream().map(Object::toString).reduce((item1, item2) -> item1 + ", " + item2).orElse(""))
              .append("]\n");
        });
      }
    }
```

and then you can probably revert the rest of the changes to this method.

This (a) only considers uniforms, (b) considers both array and non-array uniforms.

Perhaps you could adapt it to drop the [ ] if there is only one number in the list of numbers.

Comment (amakin-siru) [reply ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java]: Done.

File: ast/src/main/java/com/graphicsfuzz/common/tool/UniformValueSupplier.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright 2018 The GraphicsFuzz Project Authors
```
Comment (paulthomson) [inline ast/src/main/java/com/graphicsfuzz/common/tool/UniformValueSupplier.java:2]: Change copyright year to 2020

Comment (amakin-siru) [reply ast/src/main/java/com/graphicsfuzz/common/tool/UniformValueSupplier.java]: Changed.

**Comment by afd** [comment]
@paulthomson would you mind re-reviewing this, if @amakin-siru is indeed ready for it?

**Review by afd — CHANGES_REQUESTED** [review]
Looks good!

I had a few nit comments and also decided it would be nicer for the uniform value comments to be one per uniform, on a line each, preceding the declaration.

I think this will lead to more readable CTS tests, and it also avoids the need for special casing about semi-colongs.

This patch contains my changes.  Can you apply it and check that it looks OK?  If so, please push and I can merge.

[patch.txt](https://github.com/google/graphicsfuzz/files/5136408/patch.txt)

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 3]
```diff
@@ -172,10 +204,43 @@ public void visitDeclarationStmt(DeclarationStmt declarationStmt) {
...
+        });
+        out.append("\n");
+      }
+    }
+
     final Type baseType = variablesDeclaration.getBaseType();
     visit(baseType);
     out.append(" ");
     boolean first = true;
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:243]: Avoid adding unnecessary whitespace in a PR (unless the PR is a "code cleanup" PR).  (Apologies if this came from my patch!)

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 4]
```diff
@@ -186,6 +251,7 @@ public void visitVariablesDeclaration(VariablesDeclaration variablesDeclaration)
         out.append("[");
         visit(vdi.getArrayInfo().getSizeExpr());
         out.append("]");
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:254]: Same.

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 5]
```diff
@@ -197,6 +263,7 @@ public void visitVariablesDeclaration(VariablesDeclaration variablesDeclaration)
         visit(vdi.getInitializer());
       }
     }
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:266]: Same.

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 6]
```diff
@@ -607,6 +674,32 @@ public void visitExprCaseLabel(ExprCaseLabel exprCaseLabel) {
 
   @Override
   public void visitInterfaceBlock(InterfaceBlock interfaceBlock) {
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:677]: A comment analogous to the comment for the variables case ("// If uniform variables are being declared and we have ...") would be appropriate here - but talking about a uniform block instead of uniform variables.

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 7]
```diff
@@ -619,6 +712,7 @@ public void visitInterfaceBlock(InterfaceBlock interfaceBlock) {
     for (String memberName : interfaceBlock.getMemberNames()) {
       out.append(indent());
       final Type memberType = interfaceBlock.getMemberType(memberName).get();
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:715]: Avoid unnecessary whitespace changes.

File: ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java [hunk 8]
```diff
@@ -695,6 +789,7 @@ public String toString() {
   protected <T extends IAstNode> void visitChildFromParent(Consumer<T> visitorMethod, T child,
       IAstNode parent) {
     super.visitChildFromParent(visitorMethod, child, parent);
+
```
Comment (afd) [inline ast/src/main/java/com/graphicsfuzz/common/tool/PrettyPrinterVisitor.java:792]: Same.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 9]
```diff
@@ -70,6 +71,8 @@
 
   private final IReductionPassManager passManager;
 
+  private final Optional<IReductionPassManager> literalToUniformPassManager;
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:75]: Let's get rid of this field.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 10]
```diff
@@ -98,11 +101,14 @@ public ReductionDriver(ReducerContext context,
...
-      this.passManager = ReductionDriver.getLiteralsToUniformsPassManager(context, verbose);
+      this.literalToUniformPassManager =
+          Optional.of(ReductionDriver.getLiteralsToUniformsPassManager(context,
+          verbose));
     } else {
-      this.passManager = ReductionDriver.getDefaultPassManager(context, verbose);
+      this.literalToUniformPassManager = Optional.empty();
     }
 
+    this.passManager = ReductionDriver.getDefaultPassManager(context, verbose);
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:111]: Revert these changes - we only want one pass manager.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 11]
```diff
@@ -282,7 +288,13 @@ public String doReduction(
...
         LOGGER.info("Trying reduction attempt " + stepCount + " (" + numSuccessfulReductions
             + " successful so far).");
-        final Optional<ShaderJob> maybeNewState = passManager.applyReduction(currentState);
+        final Optional<ShaderJob> maybeNewState;
+        if (literalToUniformPassManager.isPresent()) {
+          maybeNewState = literalToUniformPassManager.get().applyReduction(currentState);
+        } else {
+          maybeNewState = passManager.applyReduction(currentState);
+        }
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:297]: Revert this - we only want one pass manager.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 12]
```diff
@@ -301,7 +313,12 @@ public String doReduction(
             makeArrayAccessesInBounds,
             addInitializers,
             currentShaderJobShortName);
-        passManager.notifyInteresting(interesting);
+        if (literalToUniformPassManager.isPresent()) {
+          literalToUniformPassManager.get().notifyInteresting(interesting);
+        } else {
+          passManager.notifyInteresting(interesting);
+        }
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:321]: Revert this - we only want one pass manager.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 13]
```diff
@@ -450,10 +467,20 @@ private void writeState(ShaderJob state, File shaderJobFileOutput,
...
-        stateToWrite,
-        shaderJobFileOutput
-    );
+
+    if (literalToUniformPassManager.isPresent()) {
+      fileOps.writeShaderJobFile(
+          stateToWrite,
+          shaderJobFileOutput,
+          Optional.of(new PipelineUniformValueSupplier(stateToWrite.getPipelineInfo()))
+      );
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:476]: Let's always do this, regardless of whether we have a "literal to uniform" pass manager.

It's always a nice idea to emit comments showing the known values of uniforms - it can only make the final shader more readable, so I figure we should do it regardless of the pass manager.

Comment (afd) [reply reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:476]: i.e., revert this change to what was there before, but pass the extra "Optional.of(...)" argument.

**Review by afd — CHANGES_REQUESTED** [review]
Getting close!

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 14]
```diff
@@ -282,7 +282,9 @@ public String doReduction(
       while (true) {
         LOGGER.info("Trying reduction attempt " + stepCount + " (" + numSuccessfulReductions
             + " successful so far).");
-        final Optional<ShaderJob> maybeNewState = passManager.applyReduction(currentState);
+        final Optional<ShaderJob> maybeNewState;
+        maybeNewState = passManager.applyReduction(currentState);
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:287]: Can you revert this to what it was previously?  No need to split this into multiple lines.

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 15]
```diff
@@ -302,6 +304,7 @@ public String doReduction(
             addInitializers,
             currentShaderJobShortName);
         passManager.notifyInteresting(interesting);
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:307]: Please avoid adding redundant whitespace: do a git diff with the last commit before your changes and get rid of any whitespace changes that are not part of the PR.

(Sometimes we do PRs whose intent is to clean up formatting, but it's best to do them separately.)

File: reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java [hunk 16]
```diff
@@ -450,9 +453,11 @@ private void writeState(ShaderJob state, File shaderJobFileOutput,
     if (addInitializers) {
       AddInitializers.addInitializers(stateToWrite);
     }
+
```
Comment (afd) [inline reducer/src/main/java/com/graphicsfuzz/reducer/ReductionDriver.java:456]: Same comment regarding whitespace.

**Review by afd — CHANGES_REQUESTED** [review]
